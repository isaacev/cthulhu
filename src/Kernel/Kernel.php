<?php

namespace Cthulhu\Kernel;

use Cthulhu\Codegen\Builder;
use Cthulhu\Codegen\PHP;
use Cthulhu\IR;
use Cthulhu\Types;

class Kernel {
  public static function IO(): IR\NativeModule {
    return self::module('IO', [
      self::fn([
        'name' => 'println',
        'signature' => new Types\FnType([ new Types\StrType() ], new Types\VoidType()),
        'stmt' => function (IR\Symbol $symbol) {
          $ref = new PHP\Reference([$symbol->name]);
          $str = new PHP\Variable('str');
          $params = [ $str ];
          $body = new PHP\BlockNode([
            new PHP\EchoStmt(
              new PHP\BinaryExpr('.',
                new PHP\VariableExpr($str),
                new PHP\StrExpr('\\n'))),
          ]);
          return new PHP\FuncStmt($ref, $params, $body);
        },
      ])
    ]);
  }

  private static function module(string $name, array $items): IR\NativeModule {
    $module = new IR\NativeModule($name);
    foreach ($items as $item) {
      switch ($item['type']) {
        case 'fn':
          $name = $item['fields']['name'];
          $symbol = new IR\Symbol($name, null, $module->scope->symbol);
          $signature = $item['fields']['signature'];
          $stmt = $item['fields']['stmt']($symbol);
          $module->fn($symbol, $signature, $stmt);
          break;
      }
    }
    return $module;
  }

  private static function fn(array $fields) {
    return [
      'type' => 'fn',
      'fields' => $fields
    ];
  }
}
