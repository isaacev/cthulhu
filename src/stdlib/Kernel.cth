--
-- NATIVE TYPES
--
-- The primitive types that are hard-coded into the language.
--

#[no_linked_kernel_types]
mod Types {
  native type Str;

  native type Float;

  native type Int;

  native type Bool;
}

--
-- PHP BUILTIN FUNCTIONS
--
-- Functions (and function-like constructs) that are provided by the PHP
-- standard library. These functions shouldn't be called from user-code but
-- should be wrapped inside of other functions in the Cthulhu standard library.
--

mod Builtins {
  #[inline]
  #[construct]
  native fn print(Str) -> ();

  #[inline]
  native fn mt_rand(Int, Int) -> Int;

  #[inline]
  #[construct]
  native fn subscript(['a], Int) -> 'a;

  #[inline]
  #[construct]
  native fn cast_int_to_string(Int) -> Str;

  #[inline]
  #[construct]
  native fn cast_float_to_string(Float) -> Str;
}
