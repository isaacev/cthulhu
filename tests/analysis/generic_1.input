use ::Kernel::*;

fn id[A](a: [A]) -> A {
  123
}

#[entry]
fn main() -> () {
  id[Int]([1, 2, 3]);
}
