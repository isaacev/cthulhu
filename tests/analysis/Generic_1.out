
  INCORRECT RETURN TYPE  -----------------------------------------------------

  The function is defined as returning the type `'a`:

    : Generic_1
  1 | fn id(a: 'a) -> 'a {
    |                 ^^

  But the function body returns the type `Int` instead:

    : Generic_1
  2 |   123
    |   ^^^

